src/hypothesis/extra/datetime.py

```diff
@@ -23,62 +32,185 @@
 
 import hypothesis.internal.conjecture.utils as cu
 from hypothesis.errors import InvalidArgument
-from hypothesis.strategies import defines_strategy
+from hypothesis._settings import note_deprecation
+from hypothesis.strategies import defines_strategy, check_valid_interval
 from hypothesis.searchstrategy.strategies import SearchStrategy
 
+__all__ = ['datetimes', 'dates', 'times']
+
+
+def convert_year_to_datetime(year, name, no_new_args=True):
+    """Conversion logic for deprecated max_year, min_year arguments."""
+    assert name in ('min_year', 'max_year')
+    assert year is not None
+    msg = ('The %s argument is deprecated.  Please use %s_datetime '
+           'with a datetime.datetime or datetime.date object instead.'
+           ) % (name, name[:3])
+    if no_new_args:
+        note_deprecation(msg)
+    else:
+        msg += '  It is an error to use new and deprecated arguments.'
+        raise InvalidArgument(msg)
+    check_valid_interval(dt.MINYEAR, year, 'dt.MINYEAR', name)
+    check_valid_interval(year, dt.MAXYEAR, name, 'dt.MAXYEAR')
+    to = dt.datetime.max if name == 'max_year' else dt.datetime.min
+    return to.replace(year=year)
+
+
+def ensure_datetime_bound(val, name):
+    """Ensure that the given value is a datetime, naive or in UTC."""
+    assert name in ('min_datetime', 'max_datetime')
+    default = dt.datetime.max if 'max_' in name else dt.datetime.min
+    if val is None:
+        return default
+    if not isinstance(val, dt.date):
+        raise InvalidArgument(
+            '%s=%r is not a date or datetime object' % (name, val))
+    if not isinstance(val, dt.datetime):
+        return default.replace(val.year, val.month, val.day)
```

Comment:
> I would also not be shocked to learn that using replace() in this way breaks in certain circumstances, although I’m unable to think of or find an example right now.

I stared at it for a while and tried some edge cases and I *think* it's fine. Agreed about keywords though.